File: Leetcode_Hard/_42. Trapping Rain Water/Solution.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        n = len(height)
        maxLeft, maxRight = [0]* n, [0]* n
        trapped_water = 0
        # Fill maxLeft array
        for i in range(1,n):
            maxLeft[i] = max(maxLeft[i-1], height[i-1])
        # Fill maxRight array
        for i in range(n - 2, 0, -1):
            maxRight[i] = max(maxRight[i+1], height[i+1])
        # At each position, the ammount of water trapped
        # is found by choosing the smaller value and reduce it
        # by the height
        for i in range(n):
            cur_water = min(maxLeft[i], maxRight[i]) - height[i]
            if (cur_water > 0):
                # Only when the cur_water value
                # is positive that we add to the 
                # total trapped water value
                trapped_water+=cur_water
        return trapped_water

File: Leetcode_Hard/_42. Trapping Rain Water/Solution.java
class Solution {
    public int trap(int[] height) {
        // Initialize the length of the array
        int n = height.length;
        // Intialize the two maxLeft and maxRight array
        int[] maxLeft = new int[n];
        int[] maxRight = new int[n];
        // Intialize the result total trapped water
        int trappedWater = 0;
        // Fill the maxLeft array
        for (int i = 1; i < n; i++ ){
            maxLeft[i] = Math.max(maxLeft[i-1], height[i-1]);
        }
        // Fill the maxRight array
        for (int i = n-2; i >= 0; i-- ){
            maxRight[i] = Math.max(maxRight[i+1], height[i+1]);
        }
        // Check the values of current water in a specific
        // position and add it to the total trapped water 
        // if it's positive
        for (int i = 0; i < n; i++){
            int curWater = Math.min(maxLeft[i], maxRight[i])- height[i];
            if (curWater > 0){
                trappedWater+=curWater;
            }
        }

        return trappedWater;

        
    }
}